### ISSUE TITLE
How to write a rule spanning multiple files?

### ISSUE THREAD

**Issue by Whathecode** [detekt/detekt#2857]
I am trying to write a rule which applies to all classes extending from a base class which has a specific annotation applied to it. These classes may be distributed across several files. How would I go about implementing this? Or, is this not an intended use case?

Things I've considered:

- When using `visit( root: KtFile )`  only one file at a time is visited. But, I need to keep track of state across multiple files.
- I've been trying to find usages of the annotated `KtKlass`, but I can't seem to figure out how (or if it is even possible) to navigate to all inheriting classes.

A gently push in the right direction would be greatly appreciated, since I'm thoroughly stuck at the moment and the documentation is somewhat lacking.

**Comment by cortinico** [comment]
> I am trying to write a rule which applies to all classes extending from a base class which has a specific annotation applied to it.

This is doable out of the box and should be pretty easy.

> These classes may be distributed across several files.

Are we talking about the subclasses that you want to report?

So to understand you're trying to inspect something like:

``` 
// File 1
annotation class DontSubclass

@DontSubclass
open class Father

// File 2
class FirstSon : Father()  // This should be reported

// File 3
class SecondSon : Father() // This should be reported
```

If this is the case, you don't need to keep track of state across multiple files.
If you can provide some example it would help us support you.

**Comment by Whathecode** [comment]
That is a good concise example. Where/how would I keep track of state, or navigate to the parent class, so I know `Father` shouldn't be extended? And, likewise for any other class annotated with `DontSubClass`.

Currently, I do not know how to do that.

**Comment by cortinico** [comment]
You should be able to achieve the desired behavior with a custom rule that will behave like this:

```kotlin
    override fun visitClassOrObject(klazz: KtClassOrObject) {
        super.visitClassOrObject(klazz)
        for (superEntry in klazz.superTypeListEntries) {
            val resolvedConstructor = superEntry.typeAsUserType
                    ?.referenceExpression
                    ?.getResolvedCall(bindingContext)
                    ?.resultingDescriptor as ClassConstructorDescriptor
            val resolvedAnnotations = resolvedConstructor.constructedClass.annotations
            if (resolvedAnnotations.any { it.fqName?.asString() == "com.example.DontSubclass" }) {
                report(
                        CodeSmell(
                                issue, Entity.from(klazz),
                                "You should not use class ${klazz.name} as it's subclassing."
                        )
                )
            }
        }
    }

```

What the rule is doing is basically:
* For every `class` or `object` declaration found
* Loop through all the listed superclasses (like all the classes and interfaces).
* Resolve the reference to the invoked constructor
* Get the reference to the constructed class from the constructor and get all the annotations for that class
* Check if any of the annotations of the class is the one you use to make your prohibited class.

**Comment by Whathecode** [comment]
Thank you very much! I think this will get me there. Worth noting that for my unit tests I did need to set the environment so that `bindingContext` is not `EMPTY`.

```
val rule = SomeRule( "SomeRule" )
val env = KtTestCompiler.createEnvironment().env
rule.compileAndLintWithContext( env, code )
```

I'll close this once I'm certain I have no related questions. To manage state across the inheritance tree I will likely introduce a custom visitor. But the main problem I was stuck at was the `?.referenceExpression?.getResolvedCall(bindingContext)?.resultingDescriptor as ClassConstructorDescriptor` so I think I'll manage to do what I want now.

**Comment by Whathecode** [comment]
I'm actually writing a rule to verify whether extending types from a base type annotated as `@Immutable` only contains val constructor parameters and properties, all of which are immutable types in turn.

Would there be interest to include such a rule in detekt, or it is too niche?

**Comment by cortinico** [comment]
> Would there be interest to include such a rule in detekt, or it is too niche?

The rule is interesting but, as you mention, it feels really niche to me. Specifically as you have to ask rule users to create an annotation that will be used during the inspection.

**Comment by Whathecode** [comment]
Yup. I think I'll take the fully qualified name of the expected annotation as a rule configuration. In case it is not specified in the linted code base and the rule is enabled, it will fail.

That said, there already seem to be rules which use fully qualified names in their rule configuration, e.g.: 

```
  UnnecessaryAbstractClass:
    active: true
    excludeAnnotatedClasses: ['dagger.Module']

  WildcardImport:
    active: true
    excludes: ['**/test/**', '**/androidTest/**', '**/commonTest/**', '**/jvmTest/**', '**/jsTest/**', '**/iosTest/**']
    excludeImports: ['java.util.*', 'kotlinx.android.synthetic.*']
```

But the difference here is of course these are very common defaults.

Although I understand it _is_ niche, I guess it should not be. :) There are already people out there believing that simply making a class a data class will make it immutable. It seems a perfect/easy candidate for static code analysis which prevents incorrect implementations where the promise of immutability is intended.

I introduced it to guarantee correct implementations of extending classes in consumers of my framework—to guarantee expected behavior in said framework—but currently have to rely on verifying immutability at runtime using reflection. Since I want extending projects to be able to add the rule in their projects, I will publish it on Maven or as a Github package which can be referred to using `detektPlugins`. Thank you for this great support for extensibility! :) I'll create an issue here once it is published to see whether there is a broader interest in this by detekt users.

**Comment by cortinico** [comment]
> But the difference here is of course these are very common defaults.

this, and ideally we also try to restrict those occurrences to really popular libraries (e.g. Dagger).

> Although I understand it _is_ niche, I guess it should not be. :) There are already people out there believing that simply making a class a data class will make it immutable. It seems a perfect/easy candidate for static code analysis which prevents incorrect implementations where the promise of immutability is intended.

I agree that we should support rules that promote immutability (open to ideas/suggestions here). On the other hand, we don't want Kotlin developers to rely on our own annotations or asking them to create one. Happy to hear other maintainer's opinion on this as well: @arturbosch @BraisGabin @schalkms 

> Since I want extending projects to be able to add the rule in their projects, I will publish it on Maven or as a Github package which can be referred to using detektPlugins

This is definitely something we can link in the README and promote 👍

**Comment by BraisGabin** [comment]
I like the idea of a rule that ensures inmutability. I'm going to open a new issue about that.

How can we help our users to discover 3rd party rules? I think that if we give more visibility to those rules more people will open their own rules too. If you have some ideas you can open a new issue so we can talk about it.

**Comment by cortinico** [comment]
> I think that if we give more visibility to those rules more people will open their own rules too.

Probably we should create a documentation page about it and collect all the 3rd party rules there

**Comment by schalkms** [comment]
Maybe we could create a new repo in the detekt group. But be aware that this also means more effort needed for maintenance.

**Comment by arturbosch** [comment]
A new documentation page sounds like less maintainance. A repo with maintained `property` files could be use for automatic discovery of rule sets :D

**Comment by schalkms** [comment]
> A new documentation page sounds like less maintainance. A repo with maintained property files could be of use for automatic discovery of rule sets :D

100 % agreed. This is IMHO the way to go. There exists just a small repo containing simple dictionary entries.

**Comment by BraisGabin** [comment]
should we move this to another issue to have track?

**Comment by schalkms** [comment]
> should we move this to another issue to have track?

Yes.
